### ISSUE TITLE
general question

### ISSUE THREAD

**Issue by laoshaw** [babelouest/ulfius#9]
I'm comparing ulfius to openwrt's uhttpd+cgi model, just curious, what's the benefit of using ulfius when you have many lightweight httpd server available(there are many) plus cgi/fcgi support, won't them do the simple task ulfius designed to do with small footprint too? or am I missing something here.

on the other hand, for simple webservice you have libonion and kore to use, how does ulfius compare to them?

Thanks,

**Comment by babelouest** [comment]
Hello,

This is a legitimate question.

Ulfius is a personal project designed to facilitate me to create webservices in good old C language, at first for my [house automation project](https://github.com/babelouest/angharad), then for [other](https://github.com/babelouest/glewlwyd) personal projects based on HTTP REST APIs.

Before implementing Ulfius, I looked at what library was available to create webservices in C, just a webservice, nothing else, and without too much dependencies that are useless to me.
In particular, I wanted an easy way to access http parameters, such as headers, query parameters, url parameters and body parameters.
I excluded webserver with cgi mdules because that was not my vision of a small webservice, but not for technical reasons. My vision of a small webservice is to have a single executable file, with or without a config file, and that's it.

3 years ago, there was not much, lots of abandonned projects, lots of work in progress, and in my memory, only a few libraries were available. I didn't see kore or libonion at that time.

I choose libmicrohttpd which was the best fit for me, but I ended up with a lot of code just to manipulate the http protocol, and I wanted to facilitate the webservice manipulation, based on what I saw in other languages, like nodejs express.

So in conclusion, the benefit is personal, it allowed me to create a webservice library for my own needs, and allowed me to understand more deeply the HTTP protocol. I published it because there may be other people who meet my needs or similar ones, and would use and improve ulfius.

**Comment by laoshaw** [comment]
Thank you for the answer!
